File: 프로그래머스/lv2/12978. 배달/배달.py
from heapq import *

def solution(N, road, K):
    INF = int(1e9)
    start = 1
    answer = 0
    distance = [INF] * (N+1)
    graph = [[] for _ in range(N+1)]
    q = []
    
    for a, b, c in road:
        graph[a].append((b, c))
        graph[b].append((a, c))
    
    distance[start] = 0
    heappush(q, (0, start))
    while q:
        dist, now = heappop(q)
                        
        if distance[now] < dist: continue
        if dist <= K: answer += 1
        
        for node, d in graph[now]:
            d += dist
            if distance[node] > d:
                distance[node] = d
                heappush(q, (d, node))

    return answer

File: 프로그래머스/lv2/12978. 배달/배달.java
import java.util.*;

class Solution {
    public int solution(int N, int[][] road, int K) {
        int answer = 0;
        int[] distance = new int[N+1];
        List<Node>[] graph = new ArrayList[N+1];
        PriorityQueue<Node> q = new PriorityQueue<>(new Comparator<Node>() {
            public int compare(Node o1, Node o2) {
                return o1.distance - o2.distance;
            }
        });
        
        for (int i=0;i < N+1;i++) {
            graph[i] = new ArrayList<>();
        }
        
        for (int[] temp: road) {
            int a = temp[0], b = temp[1], c = temp[2];
            graph[a].add(new Node(b, c));
            graph[b].add(new Node(a, c));
        }
        
        Arrays.fill(distance, Integer.MAX_VALUE);
        distance[1] = 0;

        q.add(new Node(1, 0));
        while (!q.isEmpty()) {
            Node now = q.remove();
            
            if (distance[now.num] < now.distance) continue;
            if (now.distance <= K) answer++;

            for (Node node: graph[now.num]) {
                int d = node.distance + now.distance;
                if (distance[node.num] > d) {
                    distance[node.num] = d;
                    q.add(new Node(node.num, d));
                }
            }
        }
        return answer;
    }
    
    class Node {
        int num;
        int distance;
        
        Node(int num, int distance) {
            this.num = num;
            this.distance = distance;
        }
    }
}